### PR TITLE
osd: fix the scrubber behavior on multiple preemption attempts

### PR DESCRIPTION
**Pull request by ronen-fr**
Latest scrub code creates a time window in which a specific scrub
is marked as "preempted", but future preemptions are prohibited.
Write operations handled are then blocked but not restarted on time.
    
Fixes: https://tracker.ceph.com/issues/48793

### CODE REVIEW

**Comment by ronen-fr** [comment]
jenkins retest this please

File: src/osd/pg_scrubber.cc [hunk 1]
```diff
@@ -616,6 +624,8 @@ bool PgScrubber::range_intersects_scrub(const hobject_t& start, const hobject_t&
  */
 void PgScrubber::add_delayed_scheduling()
 {
+  m_end = m_start; // not blocking any range now
```
Comment (athanatos) [inline src/osd/pg_scrubber.cc:627]: I don't understand this one.  Perhaps this should be set when we complete a chunk?

File: src/osd/pg_scrubber.cc [hunk 2]
```diff
@@ -585,13 +585,21 @@ bool PgScrubber::write_blocked_by_scrub(const hobject_t& soid)
 	   << preemption_data.is_preemptable() << " already preempted? "
 	   << preemption_data.was_preempted() << dendl;
 
+  if (preemption_data.was_preempted()) {
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/pg_scrubber.cc:588]: Fixed. Thanks. (when quick&dirty becomes very dirty...)

File: src/osd/pg_scrubber.cc (hunk 1 above)
Comment (ronen-fr) [reply src/osd/pg_scrubber.cc:623]: Yes to your question: this is what we would have done if we were able to complete the chunk. But as we were preempted:

There are two paths for preemption: 

1) locally, which means we are either in state BuildMap or in WaitReplicas:
We won't be scrubbing anything, but we still maintain the object range.
Setting 'end' to 'start' below (l. 591) frees that block. But it's also solved by the other change - we no longer block preemption attempt
(but that path is slower).

2) a replica might signal in its returning message that *it* was preempted (and will not be providing a map).
The primary may still be working on creating its own map. 
But then we notice the preemption, and while we
(unlike the original code) do not retry selecting a range immediately (we wait for all replicas, then spend time requeuing
vie PendingTimer), at least we allow multiple preemptions.
And - once we get to PendingTimer - there is no reason for us to maintain our hold on the old chunk

File: src/osd/pg_scrubber.cc [hunk 3]
```diff
@@ -585,13 +585,21 @@ bool PgScrubber::write_blocked_by_scrub(const hobject_t& soid)
...
+  }
+
   if (preemption_data.is_preemptable()) {
 
     if (!preemption_data.was_preempted()) {
       dout(10) << __func__ << " " << soid << " preempted" << dendl;
 
       // signal the preemption
       preemption_data.do_preempt();
+      m_end = m_start; // free the range we were scrubbing
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/pg_scrubber.cc]: See previous comment.

**Comment by ronen-fr** [comment]
https://pulpito.ceph.com/rfriedma-2021-01-28_20:14:16-rados-wip-ronenf-scrub-48793-distro-basic-smithi/
a small test run (filtered with 'thrash'). The one failure does not seem related. (?)

Waiting for longer tests.

File: src/osd/pg_scrubber.cc [hunk 4]
```diff
@@ -585,17 +585,21 @@ bool PgScrubber::write_blocked_by_scrub(const hobject_t& soid)
 	   << preemption_data.is_preemptable() << " already preempted? "
 	   << preemption_data.was_preempted() << dendl;
 
+  if (preemption_data.was_preempted()) {
+    // otherwise - write requests arriving while 'already preempted' is set
+    // but 'preemptable' is not - will not be allowed to continue, and will
+    // not be requeued on time.
+    return false;
```
Comment (dzafman) [inline src/osd/pg_scrubber.cc:592]: It feels dangerous to ignore m_start -> m_end blocking range, but I assume that we are guaranteeing that the current scrub chunk will be aborted.  Somewhere m_end = m_start might happen before eventually restarting that chunk or a smaller chunk at m_start.

Comment (ronen-fr) [reply src/osd/pg_scrubber.cc:592]: will verify. Thanks

File: src/osd/pg_scrubber.cc (hunk 1 above)
Comment (dzafman) [reply src/osd/pg_scrubber.cc:623]: > I don't understand this one. Perhaps this should be set when we complete a chunk?

m_start = m_end would be to complete a chunk

m_end = m_start when aborting/preempting the current chunk to stop blocking any range [m_start, m_end)

File: src/osd/pg_scrubber.cc (hunk 4 above)
Comment (ronen-fr) [reply src/osd/pg_scrubber.cc:592]: I don't see a scenario in which we have a problem with modifying 'end' in that direction (i.e. to be equal to 'start').
The current chunk will be aborted. And until that happens:
even the backend building of the local map is using its own copy of the start and end markers.

**Comment by neha-ojha** [comment]
https://pulpito.ceph.com/nojha-2021-02-01_21:31:14-rados-wip-39145-distro-basic-smithi/ - no related failures or out or order issues in this run

**Comment by neha-ojha** [comment]
@dzafman @athanatos This PR fixes a bug which causes multiple dead jobs in every rados run. If you are happy with this version, I'd like to merge it.

**Review by athanatos — APPROVED** [review]

**Review by dzafman — COMMENTED** [review]
LGTM

**Pull request merged**